core/serialization/download_xlsx_serializer.py

```diff
@@ -0,0 +1,608 @@
+from sqlalchemy import or_
+from sqlalchemy.orm import joinedload
+
+import core.db.entities as ents
+
+# isort: off
+from core.db.queries import get_programme_child_with_natural_keys_query, get_project_child_with_natural_keys_query
+
+# isort: on
+
+from core.util import ids
+
+
+def serialize_xlsx_data(programmes, programme_outcomes, projects, project_outcomes) -> dict[str, list[dict]]:
+    programme_ids = ids(programmes)
+
+    organisation_metadata = org_metadata_programme_id(programme_ids)
+    places_metadata = places_metadata_programme_id(programme_ids)
+
+    """Top level serialization of the download data."""
+    # Organisation level data
+    organisation_refs = [OrganisationSerializer(programme.organisation).to_dict() for programme in programmes]
+
+    # Programme level data
+    programme_refs = [
+        ProgrammeSerializer(
+            prog, places_metadata[prog.programme_id], organisation_metadata[prog.programme_id]
+        ).to_dict()
+        for prog in programmes
+    ]
+    programme_progresses = get_programme_progress(programme_ids, places_metadata, organisation_metadata)
+    place_details = get_place_details(programme_ids, places_metadata, organisation_metadata)
+    funding_questions = get_funding_questions(programme_ids, places_metadata, organisation_metadata)
+    programme_outcomes = [
+        OutcomeDataSerializer(
+            outcome_data,
+            places_metadata[outcome_data.programme.programme_id],
+            organisation_metadata[outcome_data.programme.programme_id],
+        ).to_dict()
+        for outcome_data in programme_outcomes
+    ]
+
+    # Project level data
+    project_ids = ids(projects)
+    project_details = get_project_details(projects, organisation_metadata, places_metadata)
+    project_progresses = get_project_progresses(project_ids, places_metadata, organisation_metadata)
+    funding = get_funding(project_ids, places_metadata, organisation_metadata)
+    funding_comments = get_funding_comments(project_ids, places_metadata, organisation_metadata)
+    private_investments = get_private_investments(project_ids, places_metadata, organisation_metadata)
+    outputs_ref = [OutputDimSerializer(output_dim).to_dict() for output_dim in ents.OutputDim.query.all()]
+    output_data = get_outputs(project_ids, places_metadata, organisation_metadata)
+    outcome_ref = [OutcomeDimSerializer(outcome_dim).to_dict() for outcome_dim in ents.OutcomeDim.query.all()]
+    project_outcomes = [
+        OutcomeDataSerializer(
+            outcome_data,
+            places_metadata[outcome_data.programme.programme_id],
+            organisation_metadata[outcome_data.programme.programme_id],
```

Comment:
tests are failing because this should be `outcome_data.project.project_id` because this is project level outcome data